src/models/restricted_pairwise_union_reach_surface.py

```diff
@@ -16,127 +16,335 @@
+        exact feasible region.  Then we uniformy sample from the truncated
+        region.  Explicitly, uniformly sample from the cube
+        {lbd: 0 <= lbd[i] <= 1 for each i} and accept the samples that satisfies
+        lbd[i] * sum(lbd) <= 1 for each i.
+
+        Args:
+          num_init:  The number of initial values to be chosen.
+        """
+        self._init_lbds = []
+        while len(self._init_lbds) < num_init:
+            lbd = np.random.rand(self._p)
+            if self._check_lbd_feasiblity(lbd):
+                self._init_lbds.append(lbd)
+
+    @classmethod
+    def _check_lbd_feasiblity(cls, lbd: List[float], tol=1e-6):
+        """Check if a choice of lbd falls in the feasible region.
+
+        This is an intermediate method used in self._uniform_initial_lbds.
+        """
+        for l in lbd:
+            if l < 0 or l * (sum(lbd) - l) > 1 + tol:
+                return False
+        return True
+
+    # The above methods set up the variables to be used in the iterative algorithm.
+    # The implementation of iterations starts from here.
+    def _step(self, lbd: List[float], i: int):
+        """Each step to update one coordinate of lbd.
 
         Args:
-          lbd: a length p vector indicating lambda_i for each pub.
+          lbd:  The current best guess of lbd.
+          i:  The index of the coordinate to be updated at this step.
 
         Returns:
-          the value of flattened a matrix.
+          The updated best guess of lbd[i].
+        """
+        us = np.array([self._compute_u(lbd, i, D) for D in self._Ds])
+        vs = np.array([self._compute_v(lbd, i, D) for D in self._Ds])
+        lbd_i_hat = np.inner(np.array(self._ys) - us, vs) / np.inner(vs, vs)
+        # The above line follows equation (4) in the algorithm description doc.
+        if lbd_i_hat < 0:
+            return 0
+        bound = self._get_feasible_bound(lbd, i)
+        if lbd_i_hat > bound:
+            return bound
+        return lbd_i_hat
+
+    @classmethod
+    def _compute_u(cls, lbd: List[float], i: int, D: np.array):
+        """Compute u following equation (2) in the algorithm description doc.
+
+        This is an intermediate method in self._step.
         """
+        u = 0
+        for j in range(len(lbd)):
+            for k in range(j + 1, len(lbd)):
+                if j != i and k != i:
+                    u += lbd[j] * lbd[k] * D[j, k]
+        return u
 
-        p = len(lbd)
-        self._a = lbd.reshape(p, 1) * lbd.reshape(1, p) - (np.eye(p) * lbd) ** 2
-        self._a = self._a.reshape(p * p, 1)
+    @classmethod
+    def _compute_v(cls, lbd: List[float], i: int, D: np.array):
+        """Compute v following equation (3) in the algorithm description doc.
 
-    def _get_constraints(self):
-        """Get constraints to be used in optimization.
+        This is an intermediate method in self._step.
+        """
+        v = 0
```

Comment:
Could this also be expressed as a matrix operation?

Reply:
will do